MANTA-5022 Manta Rebalancer: Introduce metrics

rhb2 opened the pull request.

rhb2 commented on rebalancer/src/metrics.rs:
```diff
@@ -0,0 +1,160 @@
+            port: 3020,
+            datacenter: "development".into(),
+            service: "1.rebalancer.localhost".into(),
+            server: "127.0.0.1".into(),
```
I mentioned in our conversation earlier (in MM) that I thought it was desirable to implement the functionality for reading in the needed information from a config file as a separate ticket.    And yes, I think it could just as easily be a HashMap.  I don't necessarily see one approach as better than the other per say, but this is a set of information that will always be required and I think of a HashMap as something that is used when the set of information required could vary in size.

rjloura replied:
See below.

rhb2 commented on rebalancer/src/metrics.rs:
```diff
@@ -0,0 +1,160 @@
+
+// Given the service configuration information, create (i.e. register) the
+// desired metrics with prometheus.
+pub fn register_metrics(labels: &MetricLabels) -> RegisteredMetrics {
```
Actually, now that I think about it, it's not obvious to me how using a HashMap instead of this structure is that much of a win.  I think that same argument could be used for almost any structure that we have defined in the rebalancer.  This at least gives a reader an idea of what is required when the server starts.  I guess the argument could be made that I could also enumerate those requirements in comments, but I do feel like the approach here gives an explicit guarantee about what information is required and it's type.

Now that I think about it, with a HashMap, don't all values have to be of the same time?  So to have mixed types in there, wouldn't I have to define some kind of enum to get around this?

rjloura replied:
I agree on the HashMap part. I think I got a bit turned around between RegisterMetrics and MetricLabels.  

I do however think it is worth while to add these fields to the config.json (via config-agent template).  Then register_metrics wouldn't need to take MetricLabels as an argument at all.  Or it could take an `Option<MetricLabels>` if you wanted to override them for testing or something.

rhb2 commented on rebalancer/src/metrics.rs:
```diff
@@ -0,0 +1,160 @@
+}
+
+impl RegisteredMetrics {
+    fn new(
```
Sorry, may I have to chock this up to probably not knowing about some rust syntactic sugar that you do.  Are you sayin we don't need the "new" function?  If so, I guess I would agree and I can eliminate it honestly.  I want to make sure I understand the question correctly before I do that though.

rjloura replied:
Yes, I don't think we need a `new()` method.

rhb2 commented on rebalancer/src/metrics.rs:
```diff
@@ -0,0 +1,160 @@
+    ).const_labels(const_labels.clone()), &["error"])
+    .expect("failed to register error_count counter");
+
+    RegisteredMetrics::new(
```
Indeed.  I thought about that.  Then the "register_metrics" function could walk the list of these things and, depending on the value for "type" that you have there, we would call a certain prometheus register macro (there are lots of different ones depending on the type of metric), then take whatever is returned from it and wrap it in something else (I guess probably an enum) and then store it in a vector or HashMap that could be accessed at the time that we want to bump a count on one of them.  If we did something like that, then we could eliminate the "RegisteredMetrics" structure entirely and use a HashMap for that.  Ultimately, I chose the approach I have here as a first recourse because it's not obvious to me that the amount of work that it would take to make a system *that* robust would ever really pay for itself -- at least not in the context of the rebalancer.  If you think it's worth it, I'll do it.  I wanted to anyway.  I just thought it would be received as overkill.

rjloura replied:
I think a likely next step here would be to implement metrics for the manager, in which case we could use what you propose directly instead of having to modify anything in the metrics module.

rhb2 commented on rebalancer/Cargo.toml:
```diff
@@ -23,6 +24,7 @@ joyent-rust-utils = { git = "https://github.com/joyent/rust-utils", tag = "v0.2.
 libmanta = { git = "https://github.com/joyent/rust-libmanta", tag = "v0.7.0" }
 mime = "0.3.13"
 md-5 = "0.8.0"
+prometheus = "0.5.0"
```
Can use 0.7.0.

rhb2 commented on rebalancer/src/libagent.rs:
```diff
@@ -799,13 +836,28 @@ fn process_assignment(
+        let metrics_host = config.host.clone();
+        let metrics_port = config.port;
+
+        thread::spawn(move || {
```
Done.

rhb2 commented on rebalancer/src/libagent.rs:
```diff
@@ -413,11 +418,15 @@ fn post(agent: Agent, mut state: State) -> Box<HandlerFuture> {
                 // an assignent out of the message body.
```
Yeah

rhb2 commented on rebalancer/src/libagent.rs:
```diff
@@ -413,11 +418,15 @@ fn post(agent: Agent, mut state: State) -> Box<HandlerFuture> {
                         );
+                        agent.metrics.lock().unwrap()
+                            .error_count
+                            .with_label_values(&[&e])
```
Yes.  You'd want to do it from one of the prometheus servers.

rhb2 commented:
No need to resume review yet.  Still have a couple things that I need to add.

rhb2 commented on agent/src/main.rs:
```diff
@@ -52,7 +52,7 @@ fn main() {
     // worth mentioning that this will likely also be the case for the agent
     // port.
-    Agent::run("0.0.0.0:7878");
+    Agent::run(Some("config.json"));
```
This needs to change to a config.toml, stored in /opt/smartdc/rebalancer-agent/etc.  Maybe I can add functionality to take a path (for testing purposes).  The work on the mako has already been done to render the config file and place it there though at least.

rhb2 commented:
I see there are conflicts with Cargo.lock that still have to be resolved.  I'll handle that after we agree on the rest of this.  I changed the counter_*_inc() function to take a reference to a HashMap instead of cloning.  I also use a counter_vec for the object count now too and I've changed the API of counter_vec_inv() now such that the third argument (i.e. bucket) is an Option<&str>.  This way object count can indeed be categorized by something like an action in the manager while the agent can leave it as a single stat.

rjloura commented on rebalancer/src/libagent.rs:
```diff
@@ -74,27 +103,37 @@ struct PathExtractor {
-    pub fn run(addr: &'static str) {
+    pub fn run(cfg_path: Option<&str>) {
+        let config = match cfg_path {
+            Some(c) => read_file(c),
```
can we change this to `parse_config_file` or `agent_config_from_file`?

rhb2 replied:
I think the way to go (since it's only returning a structure that's meaningful to the agent) is to move it in to the `impl` section for the agent itself.  Also, I don't think it needs to be public either.  I think I'll do that.

rjloura commented on rebalancer/src/libagent.rs:
```diff
@@ -410,14 +449,20 @@ fn post(agent: Agent, mut state: State) -> Box<HandlerFuture> {
             Ok(valid_body) => {
                 // Ceremony for parsing the information needed to create an
-                // an assignent out of the message body.
+                // an assigment out of the message body.
```
"assignment"

rjloura commented on rebalancer/src/metrics.rs:
```diff
@@ -0,0 +1,187 @@
+    MetricsCounterVec(CounterVec),
+}
+
+#[allow(irrefutable_let_patterns)]
```
where is the irrefutable `let` pattern in this?  Is it because `enum Metrics` only has `MetricsCounterVec`?

rhb2 replied:
```
  --> rebalancer/src/metrics.rs:64:13
   |
64 | /             if let Metrics::MetricsCounterVec(c) = metric {
65 | |                 // Increment the total.
66 | |                 c.with_label_values(&["total"]).inc();
67 | |
...  |
73 | |                 }
74 | |             }
   | |_____________^
```

rjloura replied:
Seems like the single field enum is why.  No explanation? 

No big deal.  I just haven't seen that warning before.

rjloura commented on rebalancer/src/metrics.rs:
```diff
@@ -0,0 +1,187 @@
+
+#[allow(irrefutable_let_patterns)]
+pub fn counter_vec_inc<S: ::std::hash::BuildHasher>(
+    metrics: &HashMap<String, Metrics, S>,
```
can we change this to `metrics_map` so that it is not confused with `metrics` module, and `Metrics` enum?

rhb2 replied:
I think I'm going to leave this the way it is.  This variable name is very commonly used all over the place in manta when passing around a collector (which is essentially what this is).

rjloura commented on rebalancer/src/metrics.rs:
```diff
@@ -0,0 +1,187 @@
+) {
+    match metrics.get(key) {
+        Some(metric) => {
+            if let Metrics::MetricsCounterVec(c) = metric {
```
I wonder if instead of creating a function for incrementing each type of metric (counter vs counter_vect, etc) we could just implement an `inc()` method on the `Metrics` enum like so:

```rs
impl Metrics {
  fn inc(&self) -> {
    match *self {
      Metrics::MetricsCounterVec(c) => {
        c.with_label_values(...).inc()
      },
      Metrics::<others> => {
         // do other inc methods
      }
    }
}
```

then the `counter_vec_inc()` function would then just be `metric_inc()` and internally call this other function that would do the right thing depending on the metric type.

rhb2 replied:
Given that counter_inc() and counter_vec_inc() take a different number of arguments, it's not obvious to me how we can have one API to rule them all for incrementing either type of counter.  Can you elaborate a bit more?  It could just be my misunderstanding of rust.

rjloura replied:
Was thinking there would be different ways to increment a given metric, but now I see that there are just different operations on a given metric.  I think what you have makes sense.

rhb2 replied:
I still don't know if I'm 100% tracking on your initial suggestion, but if you think it's possible to somehow do, it could be worth it.

rhb2 commented on rebalancer/src/metrics.rs:
```diff
@@ -112,21 +116,21 @@ pub fn register_metrics(labels: &ConfigMetrics) -> HashMap<String, Metrics> {
 
     metrics.insert(
-        "request_count".to_string(),
+        REQUEST_COUNT.to_string(),
```
After thinking about it a bit, coupled with something else that you had asked about, I think we can hit two birds with one stone.  I've created a type alias

```
pub type MetricsMap = HashMap<&'static str, Metrics>;
```

which can be used all over the place as the type instead of the HashMap itself.  I think this will reduce the likelihood of a developer confusing a variable name in this case with it's data type.  Not to mention, if, for whatever reason, we change the key type again in the future, there will not be as many maintenance points.

rjloura replied:
very nice!

rjloura commented on rebalancer/src/metrics.rs:
```diff
@@ -101,12 +103,12 @@ pub fn register_metrics(labels: &ConfigMetrics) -> HashMap<String, Metrics> {
         opts!(REQUEST_COUNT, "Total number of requests handled.")
             .const_labels(const_labels.clone()),
-        &["op"]
+        &["req"]
```
should we change the comment above as well?
> (e.g. op=GET, op=POST)

rhb2 replied:
Thanks, good catch.
